File: 50dias/cpp/40.cpp
#include <iostream>
#include <vector>
#include <iomanip>

using namespace std;

const int QUANTIDADE_PROJETOS = 10;
vector<float> orcamentoProjetos;

void LeCodigoProjeto(int& codigoProjeto) {
	bool codigoProjetoValido = false;
	do {
		cout << "Digite o codigo do projeto [1-";
		cout << QUANTIDADE_PROJETOS << "]: ";
		cin >> codigoProjeto;
		codigoProjetoValido = codigoProjeto >= 0 &&
			codigoProjeto <= QUANTIDADE_PROJETOS;
	} while (!codigoProjetoValido);
}

void LeValor(float& valor) {
	bool valorValido = false;
	do {
		cout << "Digite o valor [>0]: ";
		cin >> valor;
		valorValido = valor > 0.00;
	} while (!valorValido);
}

void LeTipo(char& tipo) {
	bool tipoValido = false;
	do {
		cout << "Digite o tipo de receita [R/D]: ";
		cin >> tipo;

		tipoValido = tipo == 'R' ||
			tipo == 'D';
	} while (!tipoValido);
}

void LeInformacoesDoUsuario(
	int& codigoProjeto,
	float& valor,
	char& tipo)
{
	LeCodigoProjeto(codigoProjeto);
	if (codigoProjeto > 0) {
		LeValor(valor);
		LeTipo(tipo);
	}
}

void RegistraInformacoes(
	int codigoProjeto,
	float valor,
	char tipo)
{
	if (tipo == 'R') {
		orcamentoProjetos[codigoProjeto - 1] += valor;
	}
	else {
		orcamentoProjetos[codigoProjeto - 1] -= valor;
	}
}

void ImprimeInformacoesProjetos() {
	for (int i = 0; i < QUANTIDADE_PROJETOS; i++) {
		cout << "---------------------" << endl;
		cout << "Projeto codigo " << i + 1 << endl;
		cout << "  Saldo: R$ " << fixed << setprecision(2);
		cout << orcamentoProjetos[i] << endl;
		cout << "---------------------" << endl;
	}
}

int main()
{
	orcamentoProjetos.resize(QUANTIDADE_PROJETOS);

	int codigoProjeto{ 0 };
	float valor{ 0.00 };
	char tipo{ ' ' };
	do {
		LeInformacoesDoUsuario(codigoProjeto, valor, tipo);
		if (codigoProjeto > 0) {
			RegistraInformacoes(codigoProjeto, valor, tipo);
		}
	} while (codigoProjeto != 0);

	ImprimeInformacoesProjetos();

	return 0;
}

File: 50dias/python/40.py
QUANTIDADE_PROJETOS = 10
orcamentoProjetos = [0] * 10

def LeCodigoProjeto():
    codigoProjeto = -1
    codigoProjetoValido = False
    while (not codigoProjetoValido):
        msg = 'Digite o codigo do projeto[1-{lim}]: '.format(lim=QUANTIDADE_PROJETOS)
        codigoProjeto = int(input(msg))
        codigoProjetoValido = 0 <= codigoProjeto <= QUANTIDADE_PROJETOS
    return codigoProjeto

def LeValor():
    valor = -1.00
    valorValido = False
    while (not valorValido):
        valor = float(input('Digite o valor [>0]: '))
        valorValido = valor > 0.00
    return valor

def LeTipo():
    tipo = ' '
    tipoValido = False
    while (not tipoValido):
        tipo = input('Digite o tipo [R/D]: ')
        tipoValido = tipo == 'R' or tipo == 'D'
    return tipo

def LeInformacoesDoUsuario():
    codigoProjeto = LeCodigoProjeto()
    valor = 0.00
    tipo = ' '
    if codigoProjeto > 0:
        valor = LeValor()
        tipo = LeTipo()
    return codigoProjeto, valor, tipo


def RegistraInformacoes(codigoProjeto, valor, tipo):
    if tipo == 'R':
        orcamentoProjetos[codigoProjeto - 1] += valor
    else:
        orcamentoProjetos[codigoProjeto - 1] -= valor

def ImprimeInformacoesProjeto():
    for i in range(QUANTIDADE_PROJETOS):
        print('--------------------------------')
        print('Projeto codigo {cod}'.format(cod=i+1))
        print('  Saldo: R$ {val:.2f}'.format(val=orcamentoProjetos[i]))
        print('--------------------------------')


def main():
    codigoProjeto = -1
    while codigoProjeto != 0:
        codigoProjeto, valor, tipo = LeInformacoesDoUsuario()
        if codigoProjeto > 0:
            RegistraInformacoes(codigoProjeto, valor, tipo)

    ImprimeInformacoesProjeto()

main()